casper/src/main/scala/coop/rchain/casper/util/rholang/RuntimeManager.scala

```diff
@@ -147,50 +135,85 @@ class RuntimeManager private (val emptyStateHash: ByteString, runtimeContainer:
     } yield (result.patterns, result.continuation.taggedCont.parBody.get.body)
   }
 
-  private def getResetRuntime(hash: StateHash) = {
-    val runtime   = runtimeContainer.take()
-    val blakeHash = Blake2b256Hash.fromByteArray(hash.toByteArray)
-    Try(runtime.space.reset(blakeHash)) match {
-      case Success(_) => runtime
-      case Failure(ex) =>
-        runtimeContainer.put(runtime)
-        throw ex
-    }
+  private def newEval(terms: Seq[Deploy], runtime: Runtime, initHash: StateHash)(
+      implicit scheduler: Scheduler): (StateHash, Seq[InternalProcessedDeploy]) = {
+    implicit val costAccountingAlg = CostAccountingAlg.unsafe[Task](CostAccount.zero)
+
+    def doEval(terms: Seq[Deploy],
+               hash: Blake2b256Hash,
+               acc: Vector[InternalProcessedDeploy]): (StateHash, Vector[InternalProcessedDeploy]) =
+      terms match {
+        case deploy +: rem =>
+          runtime.space.reset(hash)
+          val (cost, errors) = injAttempt(deploy, runtime.reducer, runtime.errorLog)
+          val newCheckpoint  = runtime.space.createCheckpoint()
```

Comment:
You create new checkpoint regardless of `errors` ? I thought that we will only do that when no error was thrown during execution of rholang program.

Reply:
@goral09 You just made me realize something. 

Why are we creating a checkpoint after each deploy? We purposely set up the rspace API so that we only have to call createCheckpoint at the end of the batch of deploys for that block.

Reply:
We always make a checkpoint because  we need to record the log from that deploy -- regardless of whether or not there are errors. This allows other nodes to reply the execution and confirm it does in fact end in error.

But notice that we only pass on the checkpoint to the next iteration if there were no errors: [see L155](https://github.com/rchain/rchain/pull/1308/files#diff-8172c898e2d53ee886672b92b2e248c9R155)

Reply:
Thanks @birchmd .

Reply:
Hmm, so this is a side-effect from grouping the logs with the deploys. I would definitely prefer just having some things run twice.

@guardbotmk3

Reply:
Is `checkpoint` operation  _expensive_ ? If so it would make sense to make them as few as possible.

Reply:
Yes it is. It requires rehashing the trie.

Reply:
So, the bigger the tuplespace the longer it will take?

Reply:
Yes. But if done correctly, something between logarithmically and linearly in the size: specifically if you don't touch half of the trie, you can reuse the untouched half's hash again without recomputing.